utils/src/shutdown.rs

```diff
@@ -0,0 +1,163 @@
+use std::sync::{
+    atomic::{AtomicBool, Ordering},
+    Arc,
+};
+
+use tokio::sync::watch;
+use tracing::warn;
+
+/// Shutdown Signal
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+#[non_exhaustive]
+pub enum Signal {
+    /// The initial state, the cluster is running.
+    Running,
+    /// Cluster shutdown signal.
+    ClusterShutdown,
+    /// Self shutdown signal.
+    SelfShutdown,
+}
+
+/// Shutdown trigger.
+#[derive(Debug, Clone)]
+pub struct Trigger {
+    /// Shutdown trigger Inner.
+    inner: Arc<TriggerInner>,
+}
+
+/// Shutdown trigger Inner.
+#[derive(Debug)]
+pub struct TriggerInner {
+    /// Sender for shutdown signal.
+    trigger: watch::Sender<Signal>,
+    /// State of mpsc channel.
+    mpsc_change_shutdown: AtomicBool,
+    /// State of sync follower daemon.
+    sync_follower_daemon_shutdown: AtomicBool,
+}
+
+impl Trigger {
+    /// Get the current state of the shutdown trigger.
+    #[inline]
+    #[must_use]
+    pub fn state(&self) -> Signal {
+        *self.inner.trigger.borrow()
+    }
+
+    /// Send the shutdown signal
+    #[inline]
+    pub fn shutdown(&self) {
+        if self.inner.trigger.send(Signal::SelfShutdown).is_err() {
+            warn!("no listener waiting for shutdown");
+        };
+    }
+
+    /// Check if the mpsc channel and sync follower daemon has been shutdown.
+    #[inline]
+    #[must_use]
+    pub fn check_partial_shutdown(&self) -> bool {
+        self.inner.mpsc_change_shutdown.load(Ordering::Relaxed)
+            && self
+                .inner
+                .sync_follower_daemon_shutdown
+                .load(Ordering::Relaxed)
+    }
+
+    /// Mark mpsc channel shutdown.
+    #[inline]
+    pub fn mark_channel_shutdown(&self) {
+        self.inner
+            .mpsc_change_shutdown
+            .store(true, Ordering::Relaxed);
+        if self.check_partial_shutdown() {
+            self.shutdown();
+        }
```

Comment:
I think it's advisable to omit this piece of code. The name of this function doesn't explicitly indicate that it will send a shutdown signal.

Reply:
Fixed